### ISSUE TITLE
Authentication and Authorization Using External Identity and Service Provider without Cognito Setup

### ISSUE THREAD

**Issue by TejasRGitHub** [awslabs/aws-dataall#872]
**Is your idea related to a problem? Please describe.**
Currently , data.all comes with Cognito IDP stack. Although, Cognito can be linked with external identity provider by using SAML exchange. This restricts relying on Cognito Userpools and also authentication is delegated through Cognito ( even for external identity providers ) . 

This restricts using data.all with Cognito. Organizations have their own setup of Identity Providers and Service Providers which data.all should be able to integrate. 

**Describe the solution you'd like**

Data.all should support use of External Identity Providers along with supporting use of service providers ( which will provide and maintain the team / group information for the user ) 

Design Proposal -

**EDIT**
 
![image](https://github.com/awslabs/aws-dataall/assets/71188245/7092cb3e-de50-4357-ac9e-55c6d578ce95)

The bottom part of the image shows the current data.all infrastructure which uses Cognito - External Identity Provider ( SAML ). for authentication. Here the tokens issued are provided by Cognito after the authentication is completed. These token are then validated in the Cognito Authorizer which provides the claims ( user profile, cognito:groups, etc ) to the GraphQL API Handler  

**EDIT**
The top part of the image separating line shows a design which will make use of External Identity Provider for authentication. Here the token will be issued by the external identity provider. These tokens cannot be validated with Cognito authorizer. For validating a custom authorizer lambda can be attached to the API Gateway. This lambda will 1. validate the token 2. create proper policy to allow execution of APIs. 

The group information i.e. the teams/groups a user belongs too can be fetched in the GraphQL API Handler itself.

**Comment by TejasRGitHub** [comment]
**Frontend Changes** 

Current data.all uses Amplify.js frontend library to setup Cognito auth endpoints. Amplify provides functions which are wrapped by the `CognitoAuthContext.js` . This file creates a Context and states which is used for maintaining user profile and login status, etc. This state is also use in the ReAuthentication Flow https://github.com/awslabs/aws-dataall/issues/571
Also, Amplify.js cannot work with Cognito user pools and needs Cognito setup for authentication, thus Amplify.js cannot be used for authentication with External Provider. 

Possible Solution - 

Use a NPM package which provides similar state and can work any third party external provider which OIDC protocol. 
`react-oidc-context` ( https://www.npmjs.com/package/react-oidc-context )can be used for authentication with external identity provider. 
Although this has a lot of state and functions similar to Amplify this alone doesn't solve the problem using the similar state like what is used by `CognitoAuthContext`.

A solution to this is wrapping the state and functions provided by `react-oidc-context` and augmenting with the states and functions similar to `CognitoAuthContext` in a nested layer like GenericAuthContext. As the states and the functions are similar the entire frontend can work without much changes. 

The react component tree will look like 
```js
<>
 ...
    <ReactOIDCContext>
        <GenericAuthContext>
           <App/>
        </GenericAuthContext>
 ...   
<>
```

Problem - This creates similar file which have to be maintained anything changes in the `CognitoAuthContext` file or the GenericAuthContext file. 

Proposed Solution - 

Create a generic context combining states and functions similar to the `CognitoAuthContext.js` file and combine Amplify and react-oidc-context in that file which will contain and  interlace with Amplify.js for authentication with Cognito and work with react-oidc-context when using custom authentication ( external Identity provider ).

Thus, the react component tree will look like 
```js
<>
 ...
    <>
        <GenericAuthContext>
           <App/>
        </GenericAuthContext>
    <>    
 ...   
<>
```
Tentative files needed to be changes. 

1. GenericAuthContext - New File 
2. CognitoAuthContext - Removing this file if not needed 
3. OidcAuthContext - This file can be used to initialize the react-oidc-context's AuthProvider ( which will initializer the user context , authentication states and functions provided by react-oidc-context) 
4. useGroups.js - As the Cognito token contain group information, and the token provided might not contain this information. This place can be used to fetch it . Solution - Create a GraphQL API call to fetch this ( `getUserForGroup`) and store this information in localStorage.  Any subsiquent calls to useGroups hook will fetch the group information from localStorage. This information will be deleted from localStorage after any logout ( either token expiry, reauth, normal logout ) 
5. useToken.js - This file will fetch tokens from Amplify function in case of Cognito or it will fetch token from react-oidc-context's user context. 


**Backend Changes** 




**Miscellenous Changes**

**Comment by TejasRGitHub** [comment]
**Backend Changes** 

Once the token is received on the frontend , this same token is sent as a part of the `Authorization` header. In the present data.all the token validation is provide by the Cognito Authorizer. 

In order for data.all to work with token issued by external identity a custom authorizer lambda (https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-use-lambda-authorizer.html) can be used. 

This lambda functions will 
1. Receive the token in the Authorization Header 
2. Validate the token with the public keys 
3. Generate a policy for allowing invocation of GraphQL, Search Api
4. If the groups claim ( like cognito:groups)  is not  provided in the token , then call service provider APIs to fetch groups / teams associated with the user.

Tentative file changes - 

api_handler - Change the way get_groups function is so that it works with the new group information provided by the custom authorizer lambda 
deploy/custom_resources/custom_authorizer - This folder will contain custom authorizer lambda file 








**Miscellenous Changes** 

1. cdk.json - Additional configs for setting up Custom Auth with External Identity Provider. 
cdk json template for Custom Auth - 
```js
 "custom_auth": {
                "provider": "",
                "url" : "",
                "client_id": "",
                "response_types": "code",
                "scopes": "openid profile"
            }
```

2. The current group / team information is provided by Cognito and there are apis which will fetch the cognito group information. These apis are stored in the `cognito_groups` folder. As these apis are structured in a way so that they only work with cognito , this code will be refactored and cognito functions will be renamed to make it generic .

**Comment by TejasRGitHub** [comment]
**Userguide Protection With External Identity Provider -** 

Currently , in data.all, the userguide is protected by Auth@Edge infrastructure (https://github.com/aws-samples/cloudfront-authorization-at-edge) . This works with Cognito. Here the tokens are passed as cookies to the edge lambda end points used by the Auth@Edge infrastructure

When using any external identity provider, a similar edge lambda can be setup which can validate the JWT token and redirect to the login page if the user is not authenticated. 

Design  - 

![image](https://github.com/awslabs/aws-dataall/assets/71188245/ac8d97f5-7b7b-49b0-ae7c-da38dd20a62e)

As the react-oidc-context doesn't store tokens and user data in cookies. There is a change in which the tokens are passed 

How to get the Token ? 
The request to user guide static page uses window.open() command with the link to the webpage. With this command, an Authorization Header information containing token cannot be passed , as the token is not passed with okta only authentication ( as currently using LocalStorage ) 
A work around this would be sending the token as a query parameter using the window.open( URL?q=TOKEN)`.
This token can then be validated .
If the token is valid - Send the user to the static content ( Status Code : 200 ) 
If the token is invalid - Send a redirection request to the login page ( Status Code : 3xx )

**Comment by TejasRGitHub** [comment]
Created PR for External IDP Support - https://github.com/awslabs/aws-dataall/pull/897

**Comment by dlpzx** [comment]
Released as part of 2.2.0
